x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/SearchableSnapshotsBlobStoreCacheMaintenanceIntegTests.java

```diff
@@ -0,0 +1,270 @@
+            if (numberOfEntriesInCache > previousNumberOfCachedEntries) {
+                final long nbEntries = numberOfEntriesInCache - previousNumberOfCachedEntries;
+                logger.info("--> mounted index [{}] has [{}] entries in cache", mountedIndex, nbEntries);
+                mountedIndices.put(mountedIndex, nbEntries);
+                mountedIndicesSettings.put(mountedIndex, getIndexSettings(mountedIndex));
+
+            } else {
+                logger.info("--> mounted index [{}] did not generate any entry in cache, skipping", mountedIndex);
+                assertAcked(client().admin().indices().prepareDelete(mountedIndex));
+            }
+
+            previousNumberOfCachedEntries = numberOfEntriesInCache;
+            i += 1;
+        }
+
+        ensureYellow(SNAPSHOT_BLOB_CACHE_INDEX);
+        refreshSystemIndex(true);
+
+        final long numberOfEntriesInCache = numberOfEntriesInCache();
+        logger.info("--> found [{}] entries in snapshot blob cache", numberOfEntriesInCache);
+        assertThat(numberOfEntriesInCache, equalTo(mountedIndices.values().stream().mapToLong(l -> l).sum()));
+
+        final List<String> indicesToDelete = randomSubsetOf(randomIntBetween(1, mountedIndices.size()), mountedIndices.keySet());
+        assertAcked(client().admin().indices().prepareDelete(indicesToDelete.toArray(String[]::new)));
+
+        final long expectedDeletedEntriesInCache = mountedIndices.entrySet()
+            .stream()
+            .filter(e -> indicesToDelete.contains(e.getKey()))
+            .mapToLong(Map.Entry::getValue)
+            .sum();
+        logger.info("--> deleting indices [{}] with [{}] entries in snapshot blob cache", indicesToDelete, expectedDeletedEntriesInCache);
+
+        assertBusy(() -> {
+            refreshSystemIndex(true);
+            assertThat(numberOfEntriesInCache(), equalTo(numberOfEntriesInCache - expectedDeletedEntriesInCache));
+
+            for (String mountedIndex : mountedIndices.keySet()) {
+                final Settings indexSettings = mountedIndicesSettings.get(mountedIndex);
+                assertHitCount(
+                    systemClient().prepareSearch(SNAPSHOT_BLOB_CACHE_INDEX)
+                        .setQuery(
+                            BlobStoreCacheMaintenanceService.buildDeleteByQuery(
+                                INDEX_NUMBER_OF_SHARDS_SETTING.get(indexSettings),
+                                SNAPSHOT_SNAPSHOT_ID_SETTING.get(indexSettings),
+                                SNAPSHOT_INDEX_ID_SETTING.get(indexSettings)
+                            )
+                        )
+                        .setSize(0)
+                        .get(),
+                    indicesToDelete.contains(mountedIndex) ? 0L : mountedIndices.get(mountedIndex)
+                );
+            }
+        });
+
+        final Set<String> remainingIndices = new HashSet<>(mountedIndices.keySet());
```

Comment:
Consider performing filtering with the Streams API

```java
mountedIndices.keySet().stream()
            .filter(Predicate.not(indicesToDelete::contains))
            .collect(Collectors.toSet())
```

Reply:
No preference here for me but this streaming versions looks readable to me :) => e0343a5a67a7e5101608e61c0955c540d05922ec